In [15]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [16]:
dataFrame = pd.read_csv('creditcard.csv')
dataFrame


Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

In [17]:
#1 -> fraudulent
#0 -> not fraudulent
dataFrame.Class.value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [18]:
#Porcentagem de fraudes
fraudes = dataFrame.Class.value_counts()[1]
nao_fraude = dataFrame.Class.value_counts()[0]
total = len(dataFrame)

print('Porcentagem de fraudes: {0:.2f}%'.format((fraudes/total)*100))
print('Porcentagem de não fraudes: {0:.2f}%'.format((nao_fraude/total)*100))

Porcentagem de fraudes: 0.17%
Porcentagem de não fraudes: 99.83%


<h4>Separando X e Y</h4>

In [19]:
from sklearn.model_selection import train_test_split

SEED = 1231
np.random.seed(SEED)

raw_train_x, raw_test_x, train_y, test_y = train_test_split(dataFrame.drop('Class', axis = 'columns'), dataFrame['Class'], test_size=0.25)

<h2>Função para imprimir os resultados</h2>

In [20]:
def imprimeResult(results):
    media = results['test_score'].mean()
    print("Resultados cross validation:",results['test_score'])
    print("Média da precisão: {0:.2f}".format(media*100),'%')

    #Acurracy fica entre a média + 2(desvio_padrao) e a média - 2(desvio_padrao)
    desvio_padrao = results['test_score'].std()
    print("Desvio padrão: {0:.2f}".format(desvio_padrao*100),'%')

    acumulo_baixo = media - 2*desvio_padrao
    acumulo_alto = media + 2*desvio_padrao
    print("Acurracy esta entre: [",acumulo_baixo*100,"---", acumulo_alto*100,"]")

<h1>Linear SVC</h1>

In [21]:
from sklearn.svm import LinearSVC
from sklearn.model_selection import cross_validate
from sklearn.model_selection import StratifiedKFold

model = LinearSVC()

cv = StratifiedKFold(n_splits=5, shuffle=True)

results = cross_validate(model, raw_train_x, train_y, cv = cv, return_train_score = False)

imprimeResult(results)

Resultados cross validation: [0.99824442 0.9984785  0.99843168 0.99833805 0.99840828]
Média da precisão: 99.84 %
Desvio padrão: 0.01 %
Acurracy esta entre: [ 99.82169369090222 --- 99.85434385503538 ]


In [22]:
from sklearn.dummy import DummyClassifier

dummy_stratified = DummyClassifier(strategy='stratified')
dummy_most_frequent = DummyClassifier(strategy='most_frequent')

dummy_stratified.fit(raw_train_x, train_y)
dummy_most_frequent.fit(raw_train_x, train_y)
print('Acurácia do modelo dummy stratified: {}%'.format(dummy_stratified.score(raw_test_x, test_y)*100))
print('Acurácia do modelo dummy most frequent: {0:.2f}%'.format(dummy_most_frequent.score(raw_test_x, test_y)*100))

Acurácia do modelo dummy stratified: 99.66152636161905%
Acurácia do modelo dummy most frequent: 99.82%


<h3>A precisão é alta mas, é possível notar que o modelo está treinando com poucas informações de fraude.</h3><br>

Ele pode sempre chutar que não é fraude e, ter uma acurácia muito boa.

<h1> É necessário realizar um treino com um campo de teste mais 'justo' (UnderSampling) </h1>
<p> Para isso, é necessário reduzir os casos 'não fraudados' para 492 também (50/50) </p> 
<p> Um problema é que o modelo irá treinar com menos dados </p>

In [23]:
#Pegando uma amostra de dados para testar o modelo

amostra_dataFrame = dataFrame.sample(frac = 1)  #frac = 1 pega todos os dados

fraude = amostra_dataFrame[amostra_dataFrame.Class == 1]
n_fraude = amostra_dataFrame[amostra_dataFrame.Class == 0][:492] #amostra de 492 casos não fraudados para o treino do modelo


In [24]:
#Juntando as amostras para o treino
dataFrame_justo = pd.concat([fraude, n_fraude])
dataFrame_justo

Time         V1         V2         V3         V4         V5  \
151006   94362.0 -26.457745  16.497472 -30.177317   8.904157 -17.892600   
189878  128595.0  -5.313774   2.664274  -4.250707   0.394707  -0.391383   
154693  102619.0  -2.488363   4.359019  -7.776410   5.364027  -1.823877   
10897    18690.0 -15.398845   7.472324 -19.026912  11.165526  -6.893856   
124176   77202.0  -0.356326   1.435305  -0.813564   1.993117   2.055878   
...          ...        ...        ...        ...        ...        ...   
220007  141979.0  -3.320020  -3.736668   2.437734  -1.148299   2.444560   
156148  107416.0  -0.519521   0.974787   1.939780  -0.277443   0.417863   
8358     11176.0   1.001175   0.391130   1.127838   2.866901  -0.318544   
77526    57089.0  -0.415945   1.279957   1.069914   0.505475   0.785861   
173870  121673.0  -0.555542   1.490180  -0.329070   1.032957   0.741357   

              V6         V7         V8        V9  ...       V21       V22  \
151006 -1.227904 -31.197329 -11.438920 -9.462573  ... -8.755698  3.460893   
189878  0.683526  -5.133671  -7.907790  0.215475  ...  8.664662 -2.716383   
154693 -2.445140  -4.964221   1.484890 -2.947899  ...  1.325672  1.021226   
10897  -2.120937 -14.913330  -0.721214 -7.175097  ... -2.444884  0.727495   
124176 -0.543579   0.487691   0.085449 -0.536352  ... -0.312863 -0.687874   
...          ...        ...        ...       ...  ...       ...       ...   
220007 -2.353851  -2.424062   0.478690 -0.076242  ...  0.235192 -0.235913   
156148  0.168126   0.374924  -0.029366  1.297772  ... -0.295212 -0.510804   
8358    0.073283  -0.107300  -0.002714  0.630733  ... -0.323506 -0.701140   
77526  -0.225729   1.255987  -0.586387 -0.048615  ... -0.058910  0.429313   
173870 -0.897616   1.138941   0.069596 -1.009842  ...  0.186690  0.742853   

             V23       V24       V25       V26       V27       V28  Amount  \
151006  0.896538  0.254836 -0.738097 -0.966564 -7.263482 -1.324884    1.00   
189878  0.483559  0.079235  0.311065  0.555544  0.176740  0.362907    1.00   
154693 -0.266476 -0.370880  0.365535  0.081372  0.184983 -0.211582    1.00   
10897  -0.345078 -0.981749  0.995271  0.816762  2.262942 -1.178063    1.00   
124176 -0.267003 -1.158480  0.271460 -0.155397  0.114328  0.101526    1.00   
...          ...       ...       ...       ...       ...       ...     ...   
220007  0.468239 -0.022403  0.714204 -0.308146  0.154117  0.191486   53.00   
156148 -0.367325 -0.857288  0.403181 -0.651802  0.092077  0.070163    2.12   
8358    0.044389  0.463128  0.318533 -0.236847 -0.029085  0.021337   56.88   
77526  -0.346143 -0.422125 -0.070193 -0.332779  0.071015 -0.165491   15.30   
173870 -0.226050  0.036337  0.094121 -0.378212  0.321072  0.166004   20.91   

        Class  
151006      1  
189878      1  
154693      1  
10897       1  
124176      1  
...       ...  
220007      0  
156148      0  
8358        0  
77526       0  
173870      0  

[984 rows x 31 columns]

984 linhas, sendo metade delas fraude e a outra metade não fraudadas.

<h1>Linear SVC com dados justos</h1>

In [25]:
from sklearn.svm import LinearSVC
from sklearn.model_selection import cross_validate
from sklearn.model_selection import StratifiedKFold

x = dataFrame_justo.drop('Class', axis = 'columns')
y = dataFrame_justo['Class']


raw_train_x, raw_test_x, train_y, test_y = train_test_split(x, y, test_size = 0.25) #stratified não é necessário

model = LinearSVC()
cv = StratifiedKFold(n_splits=5, shuffle=True)
results = cross_validate(model, raw_train_x, train_y, cv = cv, return_train_score = False)
imprimeResult(results)

Resultados cross validation: [0.5        0.62162162 0.5        0.71428571 0.72789116]
Média da precisão: 61.28 %
Desvio padrão: 9.91 %
Acurracy esta entre: [ 41.46267786712479 --- 81.08926182767206 ]


<h1>SVC() com dados justos</h1>

In [26]:
#Escalando as dimensões de análise do modelo usando pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.pipeline import Pipeline
from sklearn.model_selection import cross_validate
from sklearn.model_selection import StratifiedKFold

SEED = 301
np.random.seed(SEED)

scaler = StandardScaler()
model = SVC()

pipeline = Pipeline([('scaler', scaler), ('model', model)])

cv = StratifiedKFold(n_splits = 5, shuffle = True)

#AO invés do modelo, aqui é passado o pipeline
result = cross_validate(pipeline, x, y, cv = cv)

imprimeResult(result)

Resultados cross validation: [0.92893401 0.91370558 0.95431472 0.91878173 0.92346939]
Média da precisão: 92.78 %
Desvio padrão: 1.42 %
Acurracy esta entre: [ 89.95137984998156 --- 95.61683728458799 ]


<h1>DECISION TREE com dados justos</h1>

In [27]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_validate
from sklearn.model_selection import StratifiedKFold

model = DecisionTreeClassifier()

cv = StratifiedKFold(n_splits = 5, shuffle = True)

result = cross_validate(model, x, y, cv = cv)

imprimeResult(result)

Resultados cross validation: [0.89340102 0.88324873 0.89847716 0.85279188 0.90816327]
Média da precisão: 88.72 %
Desvio padrão: 1.90 %
Acurracy esta entre: [ 84.92204581628557 --- 92.52123606499487 ]


<h1>Dummy Classifier</h1>

In [28]:
from sklearn.dummy import DummyClassifier

dummy_stratified = DummyClassifier(strategy='stratified')
dummy_most_frequent = DummyClassifier(strategy='most_frequent')

dummy_stratified.fit(raw_train_x, train_y)
dummy_most_frequent.fit(raw_train_x, train_y)
print('Acurácia do modelo dummy stratified: {}%'.format(dummy_stratified.score(raw_test_x, test_y)*100))
print('Acurácia do modelo dummy most frequent: {0:.2f}%'.format(dummy_most_frequent.score(raw_test_x, test_y)*100))

Acurácia do modelo dummy stratified: 48.78048780487805%
Acurácia do modelo dummy most frequent: 47.97%


valores textuais na coluna pode complicar o modelo.
Falhas na base pode complicar o modelo. Tem que tratar ela antes de treinar o modelo.
    é possível apagar ou preecher com valores padrão.
É possível implementar um oversampling para resolver o problema.
Pegar toda a base e rodar no modelo.
Implementar matriz de comfusão.
Retirar o tempo do modelo..